### ISSUE TITLE
False Positive Unused Struct Field

### ISSUE THREAD

**Issue by werdl** [rust-lang/rust#121360]
I tried this code:

```rust
#[derive(Debug)]
pub struct Response {
    body: HashMap<String, Value>,
    status: u16,
    cookies: HashMap<String, String>,
    resource: String,
}

// ...

impl Response {
    // ...
    pub fn validate(&self) -> Result<(), String> {
        // ...
        let allowed_encodings = vec!["gzip", "deflate", "br", "identity"];

        if !allowed_encodings.contains(&self.body.get("encoding").unwrap_or(&Value::Null).as_str().unwrap_or("".to_string().as_str())) {
            return Err("Body encoding is not allowed".to_string());
        }

        Ok(())
    }
```

I expected to see this happen: I expected the code to compile without any warnings.

Instead, this happened: All of the struct fields (and the `validate` function) were reported as unused, when they were (I used the other struct fields also, I just omitted for brevity).

### Meta
`rustc --version --verbose`:
```
rustc 1.77.0-nightly (3cdd004e5 2023-12-29)
binary: rustc
commit-hash: 3cdd004e55c869faa2b7b25efd3becf50346e7d6
commit-date: 2023-12-29
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```

**Comment by Nilstrieb** [comment]
I don't entirely understand, I would expect this to be correct here. Could you provide a reproduction?
